### PR TITLE
Adds support to install all extensions of rust as one package

### PR DESCRIPTION
**Pull request by bkchr**
Currently, if you install the "rust" package, you only install rustc,
cargo and rust-std. With this patch, the rust package automatically
installs all packages that are listed as extension. So, you can use
"rustc --print sysroot" to build a path to the rust-src.

Solves: #16

### CODE REVIEW

File: rust-overlay.nix [hunk 1]
```diff
@@ -38,6 +38,37 @@ let
...
+  getDepsFromExtensions = pkgs: pkgname: stdenv:
+    with super.lib;
+      let
+        pkg = builtins.getAttr pkgname pkgs;
+        srcInfo = pkg.target.${hostTripleOf stdenv.system} or pkg.target."*";
+        extensions = srcInfo.extensions;
+        extensionNamesList = builtins.concatLists
+          (builtins.map (pkg: if (pkg.target == (hostTripleOf stdenv.system)) || (pkg.target == "*") then [pkg.pkg] else []) extensions);
+      in
+        unique extensionNamesList;
```
Comment (nbp) [inline rust-overlay.nix]: What is `unique` made for, is there any duplicated entry in the TOML file, or in the attribute set?

File: rust-overlay.nix [hunk 2]
```diff
@@ -38,6 +38,37 @@ let
...
   }.${system} or (throw "Rust overlay does not support ${system} yet.");
 
+  getDepsFromExtensions = pkgs: pkgname: stdenv:
+    with super.lib;
+      let
+        pkg = builtins.getAttr pkgname pkgs;
+        srcInfo = pkg.target.${hostTripleOf stdenv.system} or pkg.target."*";
+        extensions = srcInfo.extensions;
+        extensionNamesList = builtins.concatLists
+          (builtins.map (pkg: if (pkg.target == (hostTripleOf stdenv.system)) || (pkg.target == "*") then [pkg.pkg] else []) extensions);
```
Comment (nbp) [inline rust-overlay.nix]: Aren't other targets needed for cross-compilation, such as `wasm32-unknown-emscripten`?  Wouldn't it be better to keep the target architecture with it, such that we can locate the correct attribute under the top-level of the toml file?

**Comment by bkchr** [comment]
Okay, I understand why you do not want to add this as default. MakeOverridable sounds like a good solution. Tomorrow, I will look into your comments and will try to add support for MakeOverridable.

**Comment by bkchr** [comment]
I addressed all your requests and added the support to use override.

So to install "rust-src" along with rust you can use:
(rustChannels.stable.rust.override { extensions = [ "rust-src"]; })

It also prints a message if it could not find an extension.

Regarding the target architecture, I still do the comparison in the getExtensions function. But I thinkt that is okay, because you do not support cross compilation with this package at the moment?

**Comment by nbp** [comment]
@bkchr wrote:
> Regarding the target architecture, I still do the comparison in the getExtensions function. But I thinkt that is okay, because you do not support cross compilation with this package at the moment?

Indeed, I would accept your patch even without the cross-compilation story solved, as these changes move in the right direction.

I mentioned it because this is a feature requested by other persons to have feature parity with rust-nightly-nix (https://github.com/mozilla/nixpkgs-mozilla/issues/15)

**Comment by bkchr** [comment]
Okay perfect :) So, now everything should be okay or do you have any more concerns? :)

File: rust-overlay.nix [hunk 3]
```diff
@@ -38,6 +38,33 @@ let
...
+      srcInfo = pkg.target.${hostTripleOf stdenv.system} or pkg.target."*";
+    in
+      (fetchurl { url = srcInfo.url; sha256 = srcInfo.hash; });
+
+  getSrcs = pkgs: pkgname: extensions: stdenv: fetchurl:
+    with super.lib;
+      let
+        availableExtensions = getExtensions pkgs pkgname stdenv;
+        printNotFound = name: found: if found == false then builtins.trace "Extension '${name}' not found!" false else true;
+        matchedExtensions = (builtins.filter (ext: printNotFound ext (any (n: n == ext) availableExtensions)) extensions);
```
Comment (nbp) [inline rust-overlay.nix]: I see 2 issues:
 - Extension names are not discoverable without going through the manifest file.
 - builtins.trace is not preventing the user from listing a missing extension.

I think the following should solve the issue:
```nix
let
  missingExtensions = substractList availableExtensions extensions;
  installedExtensions =
    if missingExtensions == [] then extensions else throw ''
      While compiling ${pkgname}: the extension ${head missingExtensions} is not available.
      Select extensions from the following list:
      ${concatMapStringsSep (name: "${name}\n") availableExtensions}
    '';
in
```

File: rust-overlay.nix [hunk 4]
```diff
@@ -38,6 +38,33 @@ let
...
+  getFetchUrl = pkgs: pkgname: stdenv: fetchurl:
+    let
+      pkg = pkgs.${pkgname};
+      srcInfo = pkg.target.${hostTripleOf stdenv.system} or pkg.target."*";
+    in
+      (fetchurl { url = srcInfo.url; sha256 = srcInfo.hash; });
+
+  getSrcs = pkgs: pkgname: extensions: stdenv: fetchurl:
+    with super.lib;
+      let
```
Comment (nbp) [inline rust-overlay.nix]: nit: Do not indent after `with super.lib;`

**Comment by bkchr** [comment]
Fixed :)

**Comment by nbp** [comment]
After testing I realize that there is only "rust-src" listed there (on beta) and wonder if it might make sense to allow existing components to be listed in the list of extensions.  However, I think keeping this list makes sense as we would probably need it later when we start supporting different targets.

I will merge this change.
Thanks @bkchr for this contribution :)

**Comment by bkchr** [comment]
You are referring to the list of available extensions? On the stable channel I can install rust-src and rust-doc and on nightly there should also be rust-analysis.

**Pull request merged**